### ISSUE TITLE
After changing channels, I get constant errors: "Exhausted heap space, trying to allocate <N> bytes"

### ISSUE THREAD

**Issue by lukehutch** [flutter/flutter#146544]
### Steps to reproduce

I switched between Flutter channels. Now no matter which channel I use, I get constant "Exhausted heap space" errors when running any Dart code.

* `flutter channel beta` 
* `flutter upgrade && flutter pub upgrade`
* `flutter clean`
* `flutter run`

gives the following during the build:

```console
Launching lib/main.dart on sdk gphone x86 64 in debug mode...
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
Note: /home/luke/.pub-cache/hosted/pub.dev/jni-0.7.3/android/src/main/java/com/github/dart_lang/jni/JniPlugin.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
E/DartVM  ( 6389): Exhausted heap space, trying to allocate 48 bytes.
Running Gradle task 'assembleDebug'...                             37.8s
✓ Built build/app/outputs/flutter-apk/app-debug.apk
Installing build/app/outputs/flutter-apk/app-debug.apk...        1,492ms
D/FlutterImageUtilPlugin( 6653): onAttachedToEngine - IN
D/FlutterImageUtilPlugin( 6653): doOnAttachedToEngine - IN
D/FlutterImageUtilPlugin( 6653): doOnAttachedToEngine - OUT
D/FlutterImageUtilPlugin( 6653): onAttachedToEngine - OUT
D/FlutterGeolocator( 6653): Attaching Geolocator to activity
D/FlutterGeolocator( 6653): Creating service.
D/FlutterGeolocator( 6653): Binding to location service.
D/FlutterGeolocator( 6653): Geolocator foreground service connected
D/FlutterGeolocator( 6653): Initializing Geolocator services
D/FlutterGeolocator( 6653): Flutter engine connected. Connected engine count 1
I/app.clicksocial( 6653): Compiler allocated 4533KB to compile void android.view.ViewRootImpl.performTraversals()
Syncing files to device sdk gphone x86 64...                       264ms
```

and then the following errors once the app actually starts running:

```console
A Dart VM Service on sdk gphone x86 64 is available at: http://127.0.0.1:33799/FWmDyVXfOng=/
D/ProfileInstaller( 6653): Installing profile for app.clicksocial
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 32 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 176 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 176 bytes.
E/flutter ( 6653): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: ApiError
E/flutter ( 6653): ApiError
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 176 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 176 bytes.
E/flutter ( 6653): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: 
E/flutter ( 6653): 
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 32 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/flutter ( 6653): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: ApiError
E/flutter ( 6653): ApiError
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 32 bytes.
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 48 bytes.
E/flutter ( 6653): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: 
E/flutter ( 6653): 
E/DartVM  ( 6653): Exhausted heap space, trying to allocate 64 bytes.
```

The app still seems to build and launch, but these errors don't look good. The errors are both during build, and while the app is running.

Since switching Flutter channels, running anything in Dart gives me the same `Exhausted heap space` errors, for sizes between 32 bytes and about 180 bytes.

There have been similar error reports before, e.g.: #104387, #54420.

### Expected results

No error

### Actual results

Errors

### Code sample

N/A

### Screenshots or Video

N/A

### Logs

See above

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
$ flutter doctor -v
[✓] Flutter (Channel beta, 3.22.0-0.1.pre, on Fedora Linux 39 (Workstation Edition) 6.7.7-200.fc39.x86_64, locale en_US.utf8)
    • Flutter version 3.22.0-0.1.pre on channel beta at /opt/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 29babcb32a (6 days ago), 2024-04-03 17:17:04 -0500
    • Engine revision 97550907b7
    • Dart version 3.4.0 (build 3.4.0-282.1.beta)
    • DevTools version 2.34.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at /home/luke/Android/Sdk
    • Platform android-34, build-tools 33.0.1
    • Java binary at: /opt/android-studio/jbr/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 17.0.6 (Fedora 17.0.6-2.fc39)
    • cmake version 3.27.7
    • ninja version 1.11.1
    • pkg-config version 1.9.5

[✓] Android Studio (version 2022.2)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 73.0.2
    • Dart plugin version 222.4582
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.88.0)
    • VS Code at /usr/share/code
    • Flutter extension version 3.87.20240402

[✓] VS Code (version 1.88.0-insider)
    • VS Code at /usr/share/code-insiders
    • Flutter extension version 3.79.20231130

[✓] Connected device (3 available)
    • sdk gphone x86 64 (mobile) • emulator-5554 • android-x64    • Android 13 (API 33) (emulator)
    • Linux (desktop)            • linux         • linux-x64      • Fedora Linux 39 (Workstation Edition) 6.7.7-200.fc39.x86_64
    • Chrome (web)               • chrome        • web-javascript • Google Chrome 122.0.6261.94

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by lukehutch** [comment]
Actually when I run this on the emulator, I get numerous more serious errors that show that these heap allocation warnings/errors are not benign:

```console
E/flutter ( 6829): ApiError
2
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 48 bytes.
E/flutter ( 6829): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: 
E/flutter ( 6829): 
7
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 48 bytes.
E/flutter ( 6829): [ERROR:flutter/shell/common/shell.cc(117)] Dart Error: Out of memory
2
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 48 bytes.
5
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 32 bytes.
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 48 bytes.
2
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 32 bytes.
E/flutter ( 6829): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: ApiError
E/flutter ( 6829): ApiError
3
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 32 bytes.
2
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 64 bytes.
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 32 bytes.
E/flutter ( 6829): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: 
E/flutter ( 6829): 
6
E/DartVM  ( 6829): Exhausted heap space, trying to allocate 32 bytes.
```

**Comment by lukehutch** [comment]
May be related to #15479?

**Comment by lukehutch** [comment]
Switching back and forth between channels several times, and running `flutter clean` and `flutter upgrade --force` each time, seems to have somehow fixed this issue, and now I no longer see the problem anymore (so I can no longer replicate this). But I will leave the issue open in case it's useful to have a heads-up about this problem.

**Comment by huycozy** [comment]
Hi @lukehutch 
I tried reproducing this on my side with a new project (without packages) and old project (with some packages) but couldn't see the issue as reported. Does this issue only happen on your project or also can be reproduced on a new project?

Could you share the output of `flutter run -v` when issue occurs? (this could be large so you may upload it as a txt file or use online tools like https://pastebin.com/ to share it. Thanks!)

**Comment by lukehutch** [comment]
Here you go...

[flutter-run.gz](https://github.com/flutter/flutter/files/14930816/flutter-run.gz)

I killed it after a minute or so, but the errors at the end will typically continue until something like this happens (from a previous run):

```console
E/DartVM  ( 7362): Exhausted heap space, trying to allocate 32 bytes.
E/DartVM  ( 7362): Exhausted heap space, trying to allocate 32 bytes.
E/DartVM  ( 7362): Exhausted heap space, trying to allocate 48 bytes.
E/DartVM  ( 7362): Exhausted heap space, trying to allocate 176 bytes.
E/DartVM  ( 7362): Exhausted heap space, trying to allocate 48 bytes.
E/flutter ( 7362): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: ApiError
E/flutter ( 7362): ApiError
E/DartVM  ( 7362): Exhausted heap space, trying to allocate 32 bytes.
E/DartVM  ( 7362): Exhausted heap space, trying to allocate 32 bytes.
E/Dart    ( 7362): ../../third_party/dart/runtime/vm/object.cc: 2848: error: Out of memory.
E/DartVM  ( 7362): version=3.4.0-282.1.beta (beta) (Tue Apr 2 19:37:49 2024 +0000) on "android_x64"
E/DartVM  ( 7362): pid=7362, thread=7409, isolate_group=main(0x7802445a6020), isolate=main(0x7802445aa1a0)
E/DartVM  ( 7362): os=android, arch=x64, comp=yes, sim=no
E/DartVM  ( 7362): isolate_instructions=7800ae41e040, vm_instructions=7800ae41e040
E/DartVM  ( 7362): fp=7800abd18550, sp=7800abd18420, pc=7800ae229c8c
E/DartVM  ( 7362):   pc 0x00007800ae229c8c fp 0x00007800abd18550 /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!/lib/x86_64/libflutter.so+0x2382c8c
E/DartVM  ( 7362): -- End of DumpStackTrace
F/libc    ( 7362): Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 7409 (1.ui), pid 7362 (app.clicksocial)
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
Build fingerprint: 'google/sdk_gphone_x86_64/emu64xa:13/TE1A.220922.029/10205419:userdebug/dev-keys'
Revision: '0'
ABI: 'x86_64'
Timestamp: 2024-04-09 22:03:01.952891267-0600
Process uptime: 460s
Cmdline: app.clicksocial
pid: 7362, tid: 7409, name: 1.ui  >>> app.clicksocial <<<
uid: 10196
signal 6 (SIGABRT), code -1 (SI_QUEUE), fault addr --------
Abort message: '../../third_party/dart/runtime/vm/object.cc: 2848: error: Out of memory.'
    rax 0000000000000000  rbx 0000000000001cc2  rcx 00007803c05ee6ff  rdx 0000000000000006
    r8  00007800ae229c8c  r9  00007800ae229c8c  r10 00007800abd184a0  r11 0000000000000203
    r12 00007802345a6450  r13 0000000000000020  r14 00007800abd18498  r15 0000000000001cf1
    rdi 0000000000001cc2  rsi 0000000000001cf1
    rbp 0000000000000001  rsp 00007800abd18490  rip 00007803c05ee6ff
backtrace:
      #00 pc 000000000005f6ff  /apex/com.android.runtime/lib64/bionic/libc.so (abort+191) (BuildId: 007cb2313464df63debf8020e631c990)
      #01 pc 000000000226a8e8  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #02 pc 000000000230d749  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #03 pc 000000000234aa34  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #04 pc 00000000022ac769  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #05 pc 00000000025f25eb  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #06 pc 0000000002262617  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #07 pc 0000000002262562  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #08 pc 0000000002213309  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #09 pc 0000000002211eaf  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #10 pc 0000000001d92129  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #11 pc 0000000001d97726  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #12 pc 000000000001076e  /system/lib64/libutils.so (android::Looper::pollOnce(int, int*, int*, void**)+894) (BuildId: bf3f66fd6f5b7c78306a1c8270331f22)
      #13 pc 00000000000192ff  /system/lib64/libandroid.so (ALooper_pollOnce+95) (BuildId: ecffec4afc3ee31e23f9f0c32b536ba3)
      #14 pc 0000000001d976c9  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #15 pc 0000000001d9206e  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #16 pc 0000000001d95a7c  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #17 pc 0000000001d9590f  /data/app/~~w9nLR1S6pRLM_qHLuKYflg==/app.clicksocial-vWRKb3imRBmdoc8XuqsQyg==/base.apk!libflutter.so (BuildId: aeda15ebcd9d66bde8a383757158387eb361c7d4)
      #18 pc 00000000000ccd2a  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+58) (BuildId: 007cb2313464df63debf8020e631c990)
      #19 pc 0000000000060d37  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+55) (BuildId: 007cb2313464df63debf8020e631c990)
Lost connection to device.

Exited.
```

**Comment by lukehutch** [comment]
I tried a number of things, and also changed some things in my code. I can't explain how I was getting heap exhaustion errors during build originally, but I managed to get those to go away by changing flutter channels several times, and I managed to get the heap exhaustion errors to go away while running my app through code changes in my app. I suspect there was something going on here that is specific to my setup, so I will close this.
